src/typechecker.erl

```diff
@@ -3617,7 +3617,8 @@ add_type_pat({record, P, Record, Fields}, Ty, TEnv, VEnv) ->
         type_error -> throw({type_error, record_pattern, P, Record, Ty});
         {ok, Cs1} ->
             {VEnv2, Cs2} = add_type_pat_fields(Fields, Record, TEnv, VEnv),
-            {type(none), Ty, VEnv2, constraints:combine(Cs1, Cs2)}
+            RecTy = {type, P, record, [{atom, P, Record}]},
+            {type(none), RecTy, VEnv2, constraints:combine(Cs1, Cs2)}
```

Comment:
good points. (specifically the `Ty = any()` case is already handled in an earlier clause of `add_type_pat`) but GLB makes sense to me.

Reply:
Just want to be sure. Is there a test case you can think of I should aim to make sure does not fail anymore? Should this fail and remain `any()`?

```erlang
-spec any(any()) -> #test{} | not_test.
any(#test{} = Test) -> Test;
any(_) -> not_test.
```

Reply:
The above example should pass, but also the below should pass:

```Erlang
-record(test, {n :: integer()}).
-spec any(any()) -> #test{n :: 42} | not_test.
any(#test{} = Test) -> Test;
any(_) -> not_test.
```

Reply:
That test already passes as it is 🤔

Reply:
Also passes if I give a wrong value for `n`, like it being an empty list for example. Maybe that would be fixed with `glb`?

Reply:
Ah yes.., that's good! But if `any()` would be refined to `#test{}`, I think it wouldn't pass. The reason it passes now is as @gomoripeti noted: The Ty = any() case is already handled in an earlier clause of `add_type_pat`.

What else.... How about this:
```Erlang
-record(test, {n :: integer()}).
-spec test42(#test{n :: 42} | bananas) -> #test{n :: 42} | not_test.
test42(#test{} = Test) -> Test;
test42(_) -> not_test.
```

Reply:
> Also passes if I give a wrong value for `n`, like it being an empty list for example. Maybe that would be fixed with `glb`?

No, I don't think glb does anything special with this. I think it is a separate issue that we simply haven't thought of before. Thanks for finding it!

I think writing a type like `#test{n :: list()}` (given the definition of `#test{}` above) should be an error, wherever it occurs. If we allow it, we have to make sure that it's not treated as a suptype of `#test{}`, which would be a bit strange, right?

Reply:
Indeed your last does not pass, will try to fix this.

Reply:
What if RecTy is replaced by glb(Ty, RecTy) in the 2nd element in the returned tuple? I.e.

```Erlang
            RecTy = {type, erl_anno:new(0), record, [{atom, P, Record}]},
            {RefinedTy, Cs3} = glb(RecTy, Ty, TEnv),
            {type(none), RefinedTy, VEnv2, constraints:combine([Cs1, Cs2, Cs3])}
```

Does it solve the problem? We used glb for refining types in #241 so it seems reasonable to do it here as well.

Reply:
Same problem with it but thanks to you I realized I still used `P` instead of `erl_anno:new(0)` for the atom :P

Reply:
Is this thread resolved or not? I'm kind of lost.

Reply:
Since it did not solve the problems I didn't include the rest of the change as I was not sure if it was needed. I tried to create test cases that would fail with it and pass with it but didn't manage to do it. If someone can come up with a scenario where it's proven to be required or just think it can do no harm, I can include it back.

Reply:
I think there is no risk of refining `any()` to a record type here, because `any()` is handled earlier, so in that case, we don't even get here. Thus, I think we can resolve this.

Reply:
Btw, the remaining problem is #245, I assume.

Reply:
I think #245 might indeed solve it. Will resolve this one and keep track there once I open a PR.